cpmpy/tools/dimacs.py

```diff
@@ -71,42 +71,49 @@ def write_dimacs(model, fname=None):
 
 def read_dimacs(fname):
     """
-        Read a CPMpy model from a DIMACS formatted file
-        If the number of variables and constraints is not present in the header, they are inferred.
+        Read a CPMpy model from a DIMACS formatted file striclty following the specification: https://web.archive.org/web/20190325181937/https://www.satcompetition.org/2009/format-benchmarks2009.html
+        Note: the p-line has to denote the correct number of variables and clauses
         :param: fname: the name of the DIMACS file
         :param: sep: optional, separator used in the DIMACS file, will try to infer if None
     """
 
     m = cp.Model()
 
     with open(fname, "r") as f:
-
-        lines = f.readlines()
-        for i, line in enumerate(lines): # DIMACS allows for comments, skip comment lines
-            if line.startswith("p cnf"):
-                break
+        clause = []
+        had_p_line = False
+        for line in f.readlines():
+            if line == "" or line.startswith("c"):
+                continue  # skip empty and comment lines
+            elif line.startswith("p"):
+                try:
+                    _,typ,nr_vars,nr_cls = line.strip().split(" ")
+                except ValueError:
```

Comment:
Ah that's where the ValueError can come from.
Then I would parse this line in two parts:
```python
params = line.strip().split(" ")
assert len(params) == 4, f"Expected p-header to be formed as ... but got {line}"
_, typ, nr_vars, nr_cls = params
```
So without an explicit try-catch